### PR TITLE
soc/samr21: Initial support

### PR DESCRIPTION
**Pull request by benpicco**
The samr21 is an ARM Cortex-M0 part very simmilar to the samd21, but comes with a on-chip radio (AT86RF233) connected internally via SPI.

This adds support for the SoC family and the SAM R21 Xplained Pro eveluation board.
Most of it is copy & paste from `atmel_sam0/samd21`.

The AT86RF233 radio is not supported yet at this point.

### CODE REVIEW

File: boards/arm/atsamr21_xpro/doc/index.rst
Comment (dbkinder) [inline boards/arm/atsamr21_xpro/doc/index.rst]:
```suggestion
The SAM R21 Xplained Pro is a compact evaluation board by Atmel featuring a
```

File: boards/arm/atsamr21_xpro/doc/index.rst
Comment (dbkinder) [inline boards/arm/atsamr21_xpro/doc/index.rst]:
```suggestion
connected via SPI and some GPIO pins and behaves the same way as
```

File: boards/arm/atsamr21_xpro/doc/index.rst
Comment (dbkinder) [inline boards/arm/atsamr21_xpro/doc/index.rst]: This change will link to the hello world sample doc:
```suggestion
#. Build the Zephyr kernel and the :ref:`hello_world` sample application:
```

**Review by dbkinder — CHANGES_REQUESTED** [review]
a few doc changes ...

**Comment by mnkp** [comment]
This PR looks good overall, thanks @benpicco, but it needs to be split in three parts:
- ASF files provided by Atmel located in ext/
- samr21 soc files located in soc/
- atsamr21_xpro board support located in board/

The commit messages have to follow documentation guidelines [Identifying Contribution Origin](https://docs.zephyrproject.org/latest/contribute/index.html#identifying-contribution-origin). 

The header files provided by Atmel often have mixed line endings. We should change them to the Unix '\n'. Otherwise the first commit shouldn't contain any modifications to the original files.

**Review by dbkinder — APPROVED** [review]
+1 for doc changes, thanks!

File: boards/arm/atsamr21_xpro/atsamr21_xpro_defconfig
Comment (mnkp) [inline boards/arm/atsamr21_xpro/atsamr21_xpro_defconfig]: Unused code should be removed. However, board documentation claims that SPI is supported.

File: boards/arm/atsamr21_xpro/doc/index.rst
Comment (mnkp) [inline boards/arm/atsamr21_xpro/doc/index.rst]: Commit message states that "only UART, debug LED and user button are supported". Here we have also WDT, SPI.

**Review by mnkp — CHANGES_REQUESTED** [review]
There are several minor issues. Overall it looks good.

One thing which needs to be globally fixed are copyright headers. You'll need to include your name there. I'm only not sure what should happen with the name of the original author. @nashif or @galak should know.

File: boards/arm/atsamr21_xpro/atsamr21_xpro_defconfig
Comment (benpicco) [reply boards/arm/atsamr21_xpro/atsamr21_xpro_defconfig]: I could create a samr21.dtsi with
```
#include <atmel/samd21.dtsi>
&sercom4 {
       status = "ok";
       compatible = "atmel,sam0-spi";
       dipo = <0>;
       dopo = <2>;
};
```
which is the internal SPI where the radio is attached to like in [nxp_kw2xd.dtsi](https://github.com/zephyrproject-rtos/zephyr/blob/master/dts/arm/nxp/nxp_kw2xd.dtsi#L249). There a at86rf233@0 node could be attached later.

I could also just map one of the remaining SerComs to SPI on the debug header of the board.

File: boards/arm/atsamr21_xpro/doc/index.rst
Comment (benpicco) [reply boards/arm/atsamr21_xpro/doc/index.rst]: s/supported/tested/

File: soc/arm/atmel_sam0/samr21/soc.h
Comment (mnkp) [inline soc/arm/atmel_sam0/samr21/soc.h]: `ZEPHYR_ATMEL_SAMR21_SOC_H_`

File: boards/arm/atsamr21_xpro/atsamr21_xpro.dts
Comment (mnkp) [inline boards/arm/atsamr21_xpro/atsamr21_xpro.dts]: We should include `atmel/samr21.dtsi`. The "soc: atmel: add SAMR21" commit should add the required .dtsi file.

File: boards/arm/atsamr21_xpro/atsamr21_xpro_defconfig
Comment (mnkp) [reply boards/arm/atsamr21_xpro/atsamr21_xpro_defconfig]: If the code is not used it should be removed. Also - the rules are a bit fuzzy here - I believe that only basic functionality should be enabled in this file. @galak will know the definite answer.

File: soc/arm/atmel_sam0/common/soc_power.c
Comment (mnkp) [inline soc/arm/atmel_sam0/common/soc_power.c]: Let's call this function `deep_sleep`. Sleep normally means power mode where we don't loose execution context.

File: soc/arm/atmel_sam0/common/soc_power.c
Comment (mnkp) [inline soc/arm/atmel_sam0/common/soc_power.c]: I'm a bit confused myself about usage rules of leading underscore in function names, however we generally don't use them for static functions.

**Review by mnkp — COMMENTED** [review]
Power management system has been recently updated. As you probably have noticed by now this PR needs to be rebased on top of the latest master.

File: soc/arm/atmel_sam0/common/soc_power.c
Comment (benpicco) [reply soc/arm/atmel_sam0/common/soc_power.c]: I renamed the function to `sam0_idle()`

File: soc/arm/atmel_sam0/common/soc_power.c
Comment (benpicco) [reply soc/arm/atmel_sam0/common/soc_power.c]: OK, I named it `sam0_deep_sleep()`

File: soc/arm/atmel_sam0/common/Kconfig.defconfig.series
Comment (mnkp) [inline soc/arm/atmel_sam0/common/Kconfig.defconfig.series]: That's a good idea but we probably don't need a separate `Kconfig.defconfig.series` file. I would put this section directly into `soc/arm/atmel_sam0/common/Kconfig.defconfig`.

File: soc/arm/atmel_sam0/common/power.h
Comment (mnkp) [inline soc/arm/atmel_sam0/common/power.h]: This file is no longer necessary and can be removed. Functions declarations are available via `include/power.h`.

**Review by mnkp — COMMENTED** [review]
LGTM with minor comments.

Unfortunately, there is an ongoing work regarding Power Management API (#13632) which will break current implemenation. When #13632 is merged one will need to run a few string replacement operations in `soc/arm/atmel_sam0/common/power.c`.

File: soc/arm/atmel_sam0/common/Kconfig.defconfig.series
Comment (benpicco) [reply soc/arm/atmel_sam0/common/Kconfig.defconfig.series]: Ok, I was just copying what `st_stm32` was doing.

**Comment by benpicco** [comment]
I replaced the strings with their new versions from #13632

`dts_fixup.h` is copied verbatim from the samd21 folder, it's unclear who the original author was as the first git commit to it was a move from another directory.

Are Codeowners needed for files in `ext/hal`?

**Comment by benpicco** [comment]
@mnkp I replaced the strings with their new versions from #13632

The rest of the changes should be addressed now too.

File: soc/arm/atmel_sam0/samr21/Kconfig.series
Comment (benpicco) [reply soc/arm/atmel_sam0/samr21/Kconfig.series]: Ok.

File: soc/arm/atmel_sam0/samr21/Kconfig.soc
Comment (benpicco) [reply soc/arm/atmel_sam0/samr21/Kconfig.soc]: Ok.

File: soc/arm/atmel_sam0/samr21/Kconfig.soc
Comment (benpicco) [reply soc/arm/atmel_sam0/samr21/Kconfig.soc]: Ok.

File: soc/arm/atmel_sam0/samr21/CMakeLists.txt
Comment (benpicco) [reply soc/arm/atmel_sam0/samr21/CMakeLists.txt]: Ok.

File: soc/arm/atmel_sam0/samr21/Kconfig.series
Comment (benpicco) [reply soc/arm/atmel_sam0/samr21/Kconfig.series]: Removed.

File: soc/arm/atmel_sam0/samr21/soc.c
Comment (benpicco) [reply soc/arm/atmel_sam0/samr21/soc.c]: Ok.

File: soc/arm/atmel_sam0/samr21/soc.h
Comment (benpicco) [reply soc/arm/atmel_sam0/samr21/soc.h]: changed it to `ZEPHYR_ATMEL_SAMR21_SOC_H_`

File: soc/arm/atmel_sam0/common/power.h
Comment (benpicco) [reply soc/arm/atmel_sam0/common/power.h]: removed it.

File: soc/arm/atmel_sam0/samr21/soc.h
Comment (benpicco) [reply soc/arm/atmel_sam0/samr21/soc.h]: OK

File: boards/arm/atsamr21_xpro/atsamr21_xpro.dts
Comment (benpicco) [reply boards/arm/atsamr21_xpro/atsamr21_xpro.dts]: Ok.

File: boards/arm/atsamr21_xpro/atsamr21_xpro_defconfig
Comment (benpicco) [reply boards/arm/atsamr21_xpro/atsamr21_xpro_defconfig]: Ok, I removed it for now until, can add it back later when support for the radio is in place.

**Comment by benpicco** [comment]
Once #14005 #13717 and #14128 are merged they can be enabled here as well.

**Review by mnkp — COMMENTED** [review]
@benpicco Sorry for a delay, I was busy. LGTM but some things have not yet been fixed. Maybe you didn't push all the local changes you did?

Also a minor thing, the "soc: atmel: auto-select peripherals on SAM0" could be squashed with "soc: atmel: add SAMR21" as it logically belongs there.

File: soc/arm/atmel_sam0/common/power.h
Comment (benpicco) [reply soc/arm/atmel_sam0/common/power.h]: oops, sorry - I hope I didn't miss anything else, but I think here I just used `rm` instead of `git rm` and forgot to commit it.

**Comment by benpicco** [comment]
Sorry for resolving that prematurely - I'll check tomorrow if there are still any changes on my computer at work.

**Comment by benpicco** [comment]
@mnkp  Ok, I found no more changes - squashed the commit.

**Review by mnkp — APPROVED** [review]

**Comment by jfischer-phytec-iot** [comment]
@benpicco would you like to add usb support for SAM R21 Xplained board? It should be the same as for SAMD21, also pinmuxing for DP and DM.

**Comment by benpicco** [comment]
@jfischer-phytec-iot I added it to `pinmux.c` - that seems to be the only occurrence on atsamd21_xpro anyway.

**Comment by benpicco** [comment]
I sqashed the fixups.

**Comment by benpicco** [comment]
To make rebasing easier, I dropped the unrelated commits into separate pull requests.

File: soc/arm/atmel_sam0/common/soc_samd2x.c
Comment (mnkp) [inline soc/arm/atmel_sam0/common/soc_samd2x.c]: This looks like a bug in the header files. Maybe we should report it to Microchip?

File: soc/arm/atmel_sam0/samd20/CMakeLists.txt
Comment (benpicco) [reply soc/arm/atmel_sam0/samd20/CMakeLists.txt]: I've changed it so that `soc_samd2x.c` now gets selected based on whether ´CONFIG_SOC_SERIES_SAMD20`, `CONFIG_SOC_SERIES_SAMD21` or `CONFIG_SOC_SERIES_SAMR21` is set.

samd20/, samd21/ and samr21/ only contain header files and KConfig now.

File: boards/arm/atsamr21_xpro/atsamr21_xpro.yaml
Comment (jfischer-phytec-iot) [inline boards/arm/atsamr21_xpro/atsamr21_xpro.yaml]: (at least gpio and usb, i2c?)

```
supported:                                                                                                                                                                   
  - gpio                                                                                      
  - usb_device                                                                                

```

Comment (benpicco) [reply boards/arm/atsamr21_xpro/atsamr21_xpro.yaml]: I'll add gpio & usb - i2c depends on #14128

Comment (benpicco) [reply boards/arm/atsamr21_xpro/atsamr21_xpro.yaml]: Ah turns out this makes CI build some more projects with the board!
The build failure should be fixed by #14832 then. I'd rather have this merged first than manually enabling the driver implementations to the board default config.

Comment (benpicco) [reply boards/arm/atsamr21_xpro/atsamr21_xpro.yaml]: It still fails!
#15584 should fix this :crossed_fingers:

File: soc/arm/atmel_sam0/common/CMakeLists.txt
Comment (mnkp) [inline soc/arm/atmel_sam0/common/CMakeLists.txt]: Maybe we should write it as
```
zephyr_sources_ifdef(
  (CONFIG_SOC_SERIES_SAMD20 OR CONFIG_SOC_SERIES_SAMD21 OR 
   CONFIG_SOC_SERIES_SAMR21)
  soc_samd2x.c)
```

Comment (benpicco) [reply soc/arm/atmel_sam0/common/CMakeLists.txt]: If I write it like that, `soc_samd2x.c` does not get build.

**Review by nashif — APPROVED** [review]

**Pull request merged**
